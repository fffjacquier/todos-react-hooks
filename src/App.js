import React, { useReducer, useEffect, useMemo } from 'react'

import Header from './Header'
import AddTodo from './AddTodo'
import TodoList from './TodoList'
import TodoFilter from './TodoFilter'

import { fetchApiTodos } from './api'
import appReducer from './reducers'
import StateContext from './StateContext'

const App = () => {
  const [state, dispatch] = useReducer(appReducer, {
    todos: [],
    filter: 'all',
  })

  useEffect(() => {
    fetchApiTodos().then((todos) => dispatch({ type: 'FETCH_TODOS', todos }))
  }, [dispatch])

  const filteredTodos = useMemo(() => {
    const { filter, todos } = state
    switch (filter) {
      case 'active':
        return todos.filter((t) => !t.completed)
      case 'completed':
        return todos.filter((t) => t.completed)
      default:
      case 'all':
        return todos
    }
  }, [state])

  function addTodo(title) {
    dispatch({ type: 'ADD_TODO', title })
  }

  function toggleTodo(id) {
    dispatch({ type: 'TOGGLE_TODO', id })
  }

  function removeTodo(id) {
    dispatch({ type: 'REMOVE_TODO', id })
  }

  function filterTodos(filter) {
    dispatch({ type: 'FILTER_TODOS', filter })
  }

  return (
    <StateContext.Provider value={filteredTodos}>
      <div style={{ width: 400 }}>
        <Header />
        <AddTodo addTodo={addTodo} />
        <hr />
        <TodoList toggleTodo={toggleTodo} removeTodo={removeTodo} />
        <hr />
        <TodoFilter filter={state.filter} filterTodos={filterTodos} />
      </div>
    </StateContext.Provider>
  )
}

export default App
