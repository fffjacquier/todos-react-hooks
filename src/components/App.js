import React from 'react'

import Header from './Header'
import ConnectedAddTodo from '../containers/ConnectedAddTodo'
import ConnectedTodoList from '../containers/ConnectedTodoList'
import ConnectedTodoFilter from '../containers/ConnectedTodoFilter'

const App = () => {
  return (
    <div style={{ width: 400 }}>
      <Header />
      <ConnectedAddTodo />
      <hr />
      <ConnectedTodoList />
      <hr />
      <ConnectedTodoFilter />
    </div>
  )
}

export default App
