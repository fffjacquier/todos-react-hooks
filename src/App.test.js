import React from 'react'
import { unmountComponentAtNode } from 'react-dom'
//import { renderHook, act } from '@testing-library/react-hooks'
import { render, cleanup } from '@testing-library/react'
import App from './App'
//import { StateContextWrapper } from './testUtils'
//import appReducer from './reducers'

let container = null
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  // cleanup on exiting
  cleanup()
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

test('should render snapshot', () => {
  const { asFragment } = render(<App />)
  expect(asFragment(<App />)).toMatchSnapshot()
})

// test('should fetch data', () => {
//   const { result } = renderHook(() => appReducer(), {
//     wrapper: StateContextWrapper,
//   })
//   const sample = [{ id: 'test' }, { id: 'test2' }]
//   act(() => result.current.dispatch({ type: 'FETCH_TODOS', todos: sample }))
//   expect(result.current.state).toEqual(sample)
// })
