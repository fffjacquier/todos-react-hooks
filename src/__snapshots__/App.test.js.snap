// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render snapshot 1`] = `
<DocumentFragment>
  <div
    style="width: 400px;"
  >
    <h1>
      ToDos
    </h1>
    <form>
      <input
        placeholder="enter new task"
        style="width: 350px; height: 15px;"
        type="text"
        value=""
      />
      <input
        disabled=""
        style="float: right; margin-top: 2px;"
        type="submit"
        value="add"
      />
    </form>
    <hr />
    <hr />
    <div>
      <span
        style="color: blue; cursor: pointer; font-weight: bold;"
      >
        all
      </span>
       / 
      <span
        style="color: blue; cursor: pointer; font-weight: normal;"
      >
        active
      </span>
       / 
      <span
        style="color: blue; cursor: pointer; font-weight: normal;"
      >
        completed
      </span>
    </div>
  </div>
</DocumentFragment>
`;
